beamline.rst

```diff
@@ -0,0 +1,195 @@
+Beamline API
+~~~~~~~~~~~~
+
+Functionality that incorporates several different instruments or that does
+not necessarily belong to one particular instrument.
+
+
+
+**API Functions**
+
+.. code:: python
+
+   class CmdTuple(NamedTuple):
+    """
+    Describes a "generic" beamline procedure, where:
+
+    cmd_id: the string identifying the command, i.e. 'QUICK_REALIGN'
+    running: True if the procedure is running otherwise False
+    args: arguments to pass to the procedure
+    display_name: the name to display in the user interface i.e. "Quick Realign"
+    tool_tip: the tool tip
+    messages: any initial messages to display
+    """
+        cmd_id: str
+        state: CmdState
+        args: tuple
+        display_name: str
+        tool_tip: str
+        messages: list
+
+
+    def get_procedures() -> List[CmdTuple]:
+    """
+    :returns: a list of tuples each describing a beamline procedure,
+              such as: realignment, anneal
+
+    :rtype: List[CmdTuple]
+    """
+        pass
+
+        
+    def run_procedure(cmd_id:str, args) -> bool:
+    """
+    Runs procedure identified by cmd_id passes args as its arguments
+    
+    (Errors and progress that occurs is passed asynchronously via the
+     available signaling mechanism.)
+    """
+        
+    
+    def stop_procedure(cmd_id:str) -> bool:
+    """
+    Stops a running procedure identified by cmd_id
+
+    (Signal emitted if stopped or on timeout waiting for procedure to stop)
+    """
+
+
+    class Actuator(NamedTuple):
+    """
+    Describes an actuator commonly a motor, that is considered to be
+    "globally" available on the beamline. Such actuators could be shutters
+    of different kinds, i.e: fast shutter, safety shutter, other in-out
+    devices; beam stop, detector cover or even motors such as omega.
+
+    name: A unique name that identifies this actuator
+    value: The current position
+    msg: A message string, explaining state or value
+    state: The state of the actuator defined by ActuatorState
+           (MOVING, READY, ERROR ...)
+    """
+
+        name: str
+        value: float
+        msg: str
+        state: ActuatorState
+
+
+
+    def get_actuators() -> Dict[str, Actuator]:
+    """
+    :returns: A dictionary with all available actuators where the key
+              is the actuator name and the value the Actuator tuple
+    :rtype dict:            
+    """
+        pass
+
+
+    def get_actuator(name) -> Actuator:
+    """
+    :returns: The Actuator object identified by the given name
+    :rtype: Actuator
+    """
+       pass
+
+
+    def set_actuator(name, value) -> bool:
```

Comment:
Do we need a wait:bool parameter?
Or a timeout:float parameter?

Reply:
I was hesitating with the wait and timeout, however I decided that the API, that I'm suggesting, should be completely asynchronous. That is wait is always false ! and timeout, if any, determined by the underlying layer. Is there any drawbacks with that, that any once can see ?

Reply:
That sounds fine. If people wanted the UI to wait, they could then just wait till the relevant signal was sent back, right?

Reply:
Yes, that's what I had in mind, or timeout occurs and another signal is sent ...